File: typescript/src/acceptSignature.ts
import type {
  Algorithm,
  ComponentIdentifier,
  SignatureParameters,
  SignatureRequirements,
  SFVDictMember,
  SFVInnerList,
  SFVItem,
  SFVParams,
} from "./types.js";
import {
  parseDictionary,
  serializeDictionary,
  newSFVParams,
  sfvParamsSet,
  sfvParamsGet,
} from "./sfv.js";

/**
 * Build an Accept-Signature header value from a set of labeled requirements.
 *
 * Each entry maps a label (e.g. "sig1") to the requirements that the
 * corresponding signature must satisfy.
 */
export function buildAcceptSignature(
  entries: Record<string, SignatureRequirements>,
): string {
  const members: SFVDictMember[] = [];

  for (const [label, req] of Object.entries(entries)) {
    const items: SFVItem[] = req.components.map((c) => ({
      value: c.name,
      params: c.params,
    }));

    let params: SFVParams | undefined;

    if (req.keyId !== undefined) {
      if (!params) params = newSFVParams();
      params = sfvParamsSet(params, "keyid", req.keyId);
    }
    if (req.algorithm !== undefined) {
      if (!params) params = newSFVParams();
      params = sfvParamsSet(params, "alg", req.algorithm);
    }
    if (req.requireCreated) {
      if (!params) params = newSFVParams();
      params = sfvParamsSet(params, "created", true);
    }
    if (req.requireExpires) {
      if (!params) params = newSFVParams();
      params = sfvParamsSet(params, "expires", true);
    }
    if (req.tag !== undefined) {
      if (!params) params = newSFVParams();
      params = sfvParamsSet(params, "tag", req.tag);
    }

    const innerList: SFVInnerList = { items, params };
    members.push({ key: label, innerList });
  }

  return serializeDictionary(members);
}

/**
 * Parse an Accept-Signature header value into labeled requirements.
 */
export function parseAcceptSignature(
  headerValue: string,
): Record<string, SignatureRequirements> {
  const dict = parseDictionary(headerValue);
  const result: Record<string, SignatureRequirements> = {};

  for (const member of dict) {
    if (!member.innerList) continue;

    const components: ComponentIdentifier[] = member.innerList.items.map(
      (item) => {
        const name =
          typeof item.value === "string" ? item.value : String(item.value);
        return { name, params: item.params };
      },
    );

    const req: SignatureRequirements = { components };
    const params = member.innerList.params;

    if (params) {
      const keyId = sfvParamsGet(params, "keyid");
      if (typeof keyId === "string") req.keyId = keyId;

      const alg = sfvParamsGet(params, "alg");
      if (typeof alg === "string") req.algorithm = alg as Algorithm;

      const tag = sfvParamsGet(params, "tag");
      if (typeof tag === "string") req.tag = tag;

      const created = sfvParamsGet(params, "created");
      if (created === true) req.requireCreated = true;

      const expires = sfvParamsGet(params, "expires");
      if (expires === true) req.requireExpires = true;
    }

    result[member.key] = req;
  }

  return result;
}

/**
 * Convert signature requirements to SignatureParameters suitable for signing.
 *
 * The caller supplies the dynamic values (created, expires, nonce) that
 * cannot be inferred from the requirements alone.
 */
export function toSignatureParameters(
  req: SignatureRequirements,
  opts?: { created?: number; expires?: number; nonce?: string },
): SignatureParameters {
  const params: SignatureParameters = {
    components: req.components,
    keyId: req.keyId ?? "",
    algorithm: req.algorithm,
    tag: req.tag,
    created: opts?.created,
    expires: opts?.expires,
    nonce: opts?.nonce,
  };
  return params;
}

File: golang/accept_signature.go
package httpsig

import (
	"fmt"
	"sort"
)

// SignatureRequirements defines what a valid signature must contain.
// Used for both verification filtering and Accept-Signature negotiation (RFC 9421 §5).
type SignatureRequirements struct {
	// Components lists the component identifiers that must be covered by the signature.
	Components []ComponentIdentifier

	// KeyID, if non-empty, requires the signature to use this key ID.
	KeyID string

	// Algorithm, if non-empty, requires the signature to use this algorithm.
	Algorithm Algorithm

	// Tag, if non-empty, requires the signature to have this tag parameter.
	Tag string

	// RequireCreated, if true, indicates that the signature must have a created parameter.
	RequireCreated bool

	// RequireExpires, if true, indicates that the signature must have an expires parameter.
	RequireExpires bool
}

// BuildAcceptSignature serializes requirements as an Accept-Signature header value.
// Each entry maps a label to its requirements. The output is a Structured Field Values
// Dictionary where each member is an inner list of component identifiers with parameters.
func BuildAcceptSignature(entries map[string]SignatureRequirements) string {
	// Sort keys for deterministic output.
	keys := make([]string, 0, len(entries))
	for k := range entries {
		keys = append(keys, k)
	}
	sort.Strings(keys)

	members := make([]SFVDictMember, 0, len(entries))
	for _, label := range keys {
		req := entries[label]
		members = append(members, SFVDictMember{
			Key:       label,
			InnerList: requirementsToInnerList(req),
		})
	}
	return SerializeDictionary(members)
}

// requirementsToInnerList converts a SignatureRequirements to an SFVInnerList
// for serialization in the Accept-Signature header.
func requirementsToInnerList(req SignatureRequirements) *SFVInnerList {
	items := make([]SFVItem, len(req.Components))
	for i, c := range req.Components {
		items[i] = SFVItem{
			Value:  c.Name,
			Params: c.Params,
		}
	}

	params := NewSFVParams()
	if req.KeyID != "" {
		params.Set("keyid", req.KeyID)
	}
	if req.Algorithm != "" {
		params.Set("alg", string(req.Algorithm))
	}
	if req.RequireCreated {
		params.Set("created", true)
	}
	if req.RequireExpires {
		params.Set("expires", true)
	}
	if req.Tag != "" {
		params.Set("tag", req.Tag)
	}

	var p *SFVParams
	if params.Len() > 0 {
		p = params
	}

	return &SFVInnerList{
		Items:  items,
		Params: p,
	}
}

// ParseAcceptSignature parses an Accept-Signature header value into a map of
// label to SignatureRequirements. Returns an error if the header cannot be parsed.
func ParseAcceptSignature(headerValue string) (map[string]SignatureRequirements, error) {
	members, err := ParseDictionary(headerValue)
	if err != nil {
		return nil, fmt.Errorf("httpsig: malformed Accept-Signature: %w", err)
	}

	result := make(map[string]SignatureRequirements, len(members))
	for _, m := range members {
		if m.InnerList == nil {
			continue
		}

		components, err := parseComponentsFromInnerList(m.InnerList)
		if err != nil {
			return nil, fmt.Errorf("httpsig: invalid components in %q: %w", m.Key, err)
		}

		req := SignatureRequirements{
			Components: components,
		}

		if m.InnerList.Params != nil {
			if v, ok := m.InnerList.Params.Get("keyid"); ok {
				if s, ok := v.(string); ok {
					req.KeyID = s
				}
			}
			if v, ok := m.InnerList.Params.Get("alg"); ok {
				if s, ok := v.(string); ok {
					req.Algorithm = Algorithm(s)
				}
			}
			if v, ok := m.InnerList.Params.Get("tag"); ok {
				if s, ok := v.(string); ok {
					req.Tag = s
				}
			}
			if _, ok := m.InnerList.Params.Get("created"); ok {
				req.RequireCreated = true
			}
			if _, ok := m.InnerList.Params.Get("expires"); ok {
				req.RequireExpires = true
			}
		}

		result[m.Key] = req
	}

	return result, nil
}

// ToSignatureParameters converts requirements to SignatureParameters suitable for signing.
// The created, expires, and nonce values are provided by the caller since they are
// per-signature rather than per-requirement.
func (r SignatureRequirements) ToSignatureParameters(created *int64, expires *int64, nonce *string) SignatureParameters {
	// Copy components to avoid sharing the slice.
	components := make([]ComponentIdentifier, len(r.Components))
	copy(components, r.Components)

	var tag *string
	if r.Tag != "" {
		tag = StringPtr(r.Tag)
	}

	return SignatureParameters{
		Components: components,
		KeyID:      r.KeyID,
		Algorithm:  r.Algorithm,
		Created:    created,
		Expires:    expires,
		Nonce:      nonce,
		Tag:        tag,
	}
}
